client/server/isomorphic-routing/index.js

```diff
@@ -74,23 +83,38 @@ function getEnhancedContext( req, res ) {
 	} );
 }
 
-function applyMiddlewares( context, expressNext, ...middlewares ) {
-	const liftedMiddlewares = middlewares.map( middleware => next =>
-		middleware( context, err => {
-			if ( err ) {
-				expressNext( err ); // Call express' next( err ) for error handling (and bail early from this route)
-			} else {
-				next();
+function applyMiddlewares( context, ...middlewares ) {
+	// This method will run all middlewares chained. Every time a middlware calls
+	// `next`, it will pass the control to the next middleware and so on. But if
+	// at any point a middleware calls `next(error)`, only middlewares that declare
+	// 3 arguments (aka error handlers) will be called from that point.
+	const liftedMiddlewares = middlewares.map( ( middleware ) => ( next, err ) => {
+		try {
+			if ( ! err && middleware.length !== 3 ) {
+				// No errors so far, call next middleware
+				return middleware( context, next );
 			}
-		} )
-	); // prettier-ignore
+			if ( err && middleware.length === 3 ) {
+				// There is an error and this middleware can handle errors
+				return middleware( err, context, next );
+			}
+		} catch ( error ) {
+			next( error );
+		}
+
+		// At this point we are in either of these scenarios:
+		// * There is an error but this middlware is not an error handler
+		// * There is not an error but this middleware is an error handler
+		// In both cases this middleware shouldn't run, so we just skip to the next one.
+		next( err );
```

Comment:
What if we caught an `error` through `catch`? Will both `next( error )` and `next( err )` run? Is that desired?

Reply:
No, that's a bug, good catch! Only one `next(err)` should be called.